Sources/Moya/Endpoint.swift

```diff
@@ -99,7 +72,37 @@ extension Endpoint {
         request.httpMethod = method.rawValue
         request.allHTTPHeaderFields = httpHeaderFields
 
-        return try? parameterEncoding.encode(request, with: parameters)
+        switch task {
+        case .requestPlain, .uploadFile, .uploadMultipart, .downloadDestination:
+            break
+
+        case .requestData(let data):
+            request.httpBody = data
+
+        case let .requestParameters(parameters: parameters, encoding: parameterEncoding):
+            return try? parameterEncoding.encode(request, with: parameters)
+
+        case let .requestCompositeData(bodyData: bodyData, urlParameters: urlParameters):
+            do {
```

Comment:
Well, I can only use `try?` if the `request` is not needed after that line like in those other cases. But here for example the `request` type is needed to set the http body afterwards. Of course I could change the order here and set the http body first ... but on the composite parameters case I need to do two `try`s anyway so I simply kept the `do-catch`.

But since you seem not to like the `do-catch`s I've changed them up to this:

```swift
case let .requestCompositeData(bodyData: bodyData, urlParameters: urlParameters):
            request.httpBody = bodyData
            return try? URLEncoding.default.encode(request, with: urlParameters)

        case let .requestCompositeParameters(bodyParameters: bodyParameters, bodyEncoding: bodyParameterEncoding, urlParameters: urlParameters):
            guard let bodyfulRequest = try? bodyParameterEncoding.encode(request, with: bodyParameters) else { return nil }
            return try? URLEncoding.default.encode(bodyfulRequest, with: urlParameters)
```

I hope this is better.